File: Baekjoon/이분 탐색/MinjuCho/1920.py
# 문제 풀이 링크: https://velog.io/@cuppizza/백준-1920-수-찾기-파이썬-C
# 실행 시간: 268ms 메모리: 48772 KB
# 개선 풀이
from bisect import bisect_left
N = int(input())
arr = list(map(int, input().split()))
arr.sort()

M = int(input())
output = list(map(int, input().split()))
for i in range(M):
    find_pos = bisect_left(arr, output[i])
    if find_pos == N or (find_pos != N and arr[find_pos] != output[i]):
        output[i] = 0
    else:
        output[i] = 1
    print(output[i])

# 기존 풀이
# 실행 시간: 452ms 메모리: 48780 KB
def binary_search(key):
    start = 0
    end = len(arr)-1
    while start<=end:
        mid = (start + end) // 2
        if arr[mid] == key:
            return 1
        elif arr[mid] < key:
            start = mid+1
        else:
            end = mid-1
    return 0

N = int(input())
arr = list(map(int, input().split()))
arr.sort()

M = int(input())
output = list(map(int, input().split()))
for i in range(M):
    output[i] = binary_search(output[i])
    print(output[i], end = ' ')

File: Baekjoon/이분 탐색/MinjuCho/1920.cpp
// 문제 풀이 링크: https://velog.io/@cuppizza/백준-1920-수-찾기-파이썬-C
// 실행 시간: 72ms 메모리: 2012KB
#include<stdio.h>
#include<vector>
#include<algorithm>

using namespace std;
int main(void)
{
	int N, M;
	scanf_s("%d", &N);
	vector<int> arr(N);
	for (int i = 0; i < N; ++i)
	{
		scanf_s("%d", &arr[i]);
	}
	sort(arr.begin(), arr.end());
	scanf_s("%d", &M);
	vector<int> output(M);
	for (int i = 0; i < M; ++i)
	{
		int input;
		scanf_s("%d", &input);
		auto iter = lower_bound(arr.begin(), arr.end(), input);
		if ( iter!= arr.end() && *iter == input )
		{
			output[i] = 1;
		}
		else
		{
			output[i] = 0;
		}
	}
	for (int i = 0; i < M; ++i)
	{
		printf("%d\n", output[i]);
	}
}